impr(optimization/refactor): Optimized Debian dockerfiles, Enhanced readability + Enhanced code structure. (@Ilolm)

ilolm opened the pull request:
### Description

1.Reduced layeres amount.
2.Enhanced readibility.
3.Enhanced dockerfile sctructure(Moved variable in one place instead of being scattered on the rest of the code). 3.Fixed some bugs such as lowercase 'as'.

<!-- Please describe your pull request here. -->

### Testing done

<!-- Comment:
Provide a clear description of how this change was tested.
At minimum this should include proof that a computer has executed the changed lines.
Ideally this should include an automated test or an explanation as to why this change has no tests.
Note that automated test coverage is less than complete, so a successful PR build does not necessarily imply that a computer has executed the changed lines.
If automated test coverage does not exist for the lines you are changing, you must describe the scenario(s) in which you manually tested the change.
For frontend changes, include screenshots of the relevant page(s) before and after the change.
For refactoring and code cleanup changes, exercise the code before and after the change and verify the behavior remains the same.
-->

**1. Manually built an edited Dockerfile image successfully.**
```bash
docker image built --no-cache -t test_feature .
```
**2. Ran the image successfully and checked the web ui.**
```bash
docker container run --rm -p 8080:8080 -p 50000:50000 test_feature
```  

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

<!--
Put an `x` into the [ ] to show you have filled the information.
The template comes from https://github.com/jenkinsci/.github/blob/master/.github/pull_request_template.md 
You can override it by creating .github/pull_request_template.md in your own repository 
-->

dduportal reviewed (changes requested):
Thanks for this proposal!

A first pass of superficial review:

- This PR does a LOT of things. It's really hard, as a maintainer committed to have a safe and sustainable image, to review and evaluate all risks. It's not immediate which change solves which problem. Do you mind splitting in different PRs, each one solving a single problem (so easier to review, faster to merge when no discussion is needed)? Or at least open distinct issues explaining the problem and the solving strategy so we could relate and discuss (instead of having what is going to be a really big PR slow to review and merge with tons of discussions)?

- Grouping `ARG` and `ENV` does not change the layering at all, because it's only a set of metadata (docker build does not create any layer with these instructions). Besides, it introduces big downsides in this repository:
  - It breaks our [updatecli]()https://www.updatecli.io/ [manifests](https://www.updatecli.io/) system which tracks values => can you change it to, at least, distinct instructions (unless it solves a particular issue)?
  - It breaks the Docker caching when we update only a couple of "ARG + RUN" such as the jenkins-plugin cli. The "scattering" is sometimes a strong functional feature in Dockerfile to benefit from layer caching, while grouping on a single location is a syntactic sugar for code style. => Is there an issue you want to solve other than code style?"

- Moving the `LABEL` on top is breaking the values (as it uses ARG values which are defined later in the scopes).

- Mixed feelings about the `COPY` instruction in its array form. It does not reduce layers in the context of docker buildx, but it simplifies the copy. It feels a source of confusion as it requires higher skills in Linux + Docker skills. I'm torn between the simplification vs. making it harder to read and contribute

Some changes make a lot of sense:

- Cleaning up apt after the git-lfs install
- Having the VOLUME definition as last element makes sense because it avoid mistakes when writing / changing the associated mountpoint
- Grouping EXPOSE instructions
- Fixing the lower case keywords (not bugs: it's working as expected, but since Docker now prints warning about this, let's fix it)
-  Moving USER to the end to avoid errors with permissions


Some changes does not make sense for me (as individual maintainer: noit a general rule):
- Adding `echo` instructions in the `RUN` instructions is a Dockerfile code smell. it means you want to locate which instruction causes failure (when failing to build). It means you're trying to create a big layer with a LOT of moving pieces, which defeats the purpose of container images caching. Besides, the "echo" adds instruction making it heavier to read and understand.
- Adding executable bit on copied shell script should be avoided in favor of calling the shell directly with the script path as argument (personal preference in the context of Dockerfile in which we control the shell so shebang should not matter)
- The comments should be turned in to documentation: they are not following homogeneous convention and are repeating the code without providing any information

ilolm commented:
Hi there! 
Firstly I highly appreciate your review and notices made about the changes.

---

> This PR does a LOT of things. It's really hard, as a maintainer committed to have a safe and sustainable image, to review and evaluate all risks. It's not immediate which change solves which problem. Do you mind splitting in different PRs, each one solving a single problem (so easier to review, faster to merge when no discussion is needed)? Or at least open distinct issues explaining the problem and the solving strategy so we could relate and discuss (instead of having what is going to be a really big PR slow to review and merge with tons of discussions)?

Okay, I will take this into account.

---

> Grouping `ARG` and `ENV` does not change the layering at all, because it's only a set of metadata (docker build does not create any layer with these instructions). Besides, it introduces big downsides in this repository:

The main purpose of this was to made the code structure more readable by moving each piece of code into his own section. However if it breaks the code or that was made specifically for caching purposes, I will modify those things back in my repo.

---

> Moving the `LABEL` on top is breaking the values (as it uses ARG values which are defined later in the scopes).

Yeah, that's definitely my bad, sorry about that. It's gonna be fixed as well.

---

> Mixed feelings about the `COPY` instruction in its array form. It does not reduce layers in the context of docker buildx, but it simplifies the copy. It feels a source of confusion as it requires higher skills in Linux + Docker skills. I'm torn between the simplification vs. making it harder to read and contribute

As for me it looks really self-explanatory but if you want, I can change it back as well :)

---

I also agree with your personnel notices --> so, there are gonna be fixed as well. 

---

> Is there a reason to have done these changes only on the debian images?

No, I was going to make the same task with the alpine images later. 

---

> The linter (hadolint) fails (as per the CI checks) which is a prerequisite

Yeah, I see. I'm gonna fix it as well

ilolm commented:
Hadolint fix in progress

ilolm commented:
## Summarizing changes:

- Changed 'as' lowercase to 'AS' uppercase.
- Combined COPY statements into an array.
- Added 'Cleaning up' sections in RUN statements containing apt-get operations.
- Moved LABEL statement on the top with some required variables above.
- Moved VOLUME in the end.
-  Moved USER in the end before ENTRYPOINT.
- Combined EXPOSE statements and moved in the end.
- Modified comments.

dduportal commented:
Thanks @ilolm ! I'm a bit busy these time, so expect some delay in the review

timscharfenort8 reviewed (commented):
Gute Arbeit

dduportal commented:
Hi @ilolm , sorry for the delay.

I'm gonna close this PR because it embeds too many changes creating too many risks: it should be a split in a myriad of changes to make it easier to review without blocking each others.

If you are still willing to propose these linting improvement, do you mind starting with an issue listing the linter messages, so we could map each PR to a single linter message please?
If you are not willing to continue, we also understand.
